pkg/domain/cloudproviders.go

```diff
@@ -27,3 +83,112 @@ type GetCloudKeystoreRequest struct {
 	CloudKeystoreID   *string
 	CloudKeystoreName *string
 }
+
+type MachineIdentityStatus int
+
+const (
+	MachineIdentityStatusUnknown MachineIdentityStatus = iota
+	MachineIdentityStatusNew
+	MachineIdentityStatusPending
+	MachineIdentityStatusInstalled
+	MachineIdentityStatusDiscovered
+	MachineIdentityStatusValidated
+	MachineIdentityStatusMissing
+	MachineIdentityStatusFailed
+
+	MachineIdentityStatusUnknownStr    = "UNKNOWN"
+	MachineIdentityStatusNewStr        = "NEW"
+	MachineIdentityStatusPendingStr    = "PENDING"
+	MachineIdentityStatusInstalledStr  = "INSTALLED"
+	MachineIdentityStatusDiscoveredStr = "DISCOVERED"
+	MachineIdentityStatusValidatedStr  = "VALIDATED"
+	MachineIdentityStatusMissingStr    = "MISSING"
+	MachineIdentityStatusFailedStr     = "FAILED"
+
+	CloudMetadataACM     = "ACM"
+	CloudMetadataGCM     = "GCM"
+	CloudMetadataAKV     = "AKV"
+	CloudMetadataUnknown = "UNKNOWN"
+)
+
+func (mis MachineIdentityStatus) String() string {
+	switch mis {
+	case MachineIdentityStatusNew:
+		return MachineIdentityStatusNewStr
+	case MachineIdentityStatusPending:
+		return MachineIdentityStatusPendingStr
+	case MachineIdentityStatusInstalled:
+		return MachineIdentityStatusInstalledStr
+	case MachineIdentityStatusDiscovered:
+		return MachineIdentityStatusDiscoveredStr
+	case MachineIdentityStatusValidated:
+		return MachineIdentityStatusValidatedStr
+	case MachineIdentityStatusMissing:
+		return MachineIdentityStatusMissingStr
+	case MachineIdentityStatusFailed:
+		return MachineIdentityStatusFailedStr
+	default:
+		return MachineIdentityStatusUnknownStr
+	}
+}
+
+type CertificateCloudMetadata struct {
+	values map[string]interface{}
+}
+
+func NewCertificateCloudMetadata(values map[string]interface{}) CertificateCloudMetadata {
+	return CertificateCloudMetadata{
+		values: values,
+	}
+}
+
+func (ccm *CertificateCloudMetadata) GetType() string {
+	typ := ccm.GetValue("__typename")
+	if typ == nil {
+		return CloudMetadataUnknown
+	}
+	switch typ {
+	case "AWSCertificateMetadata":
+		return CloudMetadataACM
+	case "AzureCertificateMetadata":
+		return CloudMetadataGCM
+	case "GCPCertificateMetadata":
+		return CloudMetadataAKV
+	default:
+		return CloudMetadataUnknown
+	}
+}
+
+func (ccm *CertificateCloudMetadata) GetMetadata() map[string]interface{} {
+	return ccm.values
+}
+
+func (ccm *CertificateCloudMetadata) GetValue(key string) interface{} {
+	if key == "" {
+		return nil
+	}
+	if ccm.values == nil {
+		return nil
+	}
+	return ccm.values[key]
+}
+
+type CloudMachineIdentity struct {
+	ID                uuid.UUID
+	CloudKeystoreID   uuid.UUID
+	CloudKeystoreName string
+	CloudProviderID   uuid.UUID
+	CloudProviderName string
+	CertificateID     uuid.UUID
+	Metadata          *CertificateCloudMetadata
+	Status            MachineIdentityStatus
+	StatusDetails     string
+}
+
+type GetCloudMachineIdentityRequest struct {
+	KeystoreID        *string
+	MachineIdentityID *string
+	Fingerprints      []string
```

Comment:
If MI ID is mandatory, then why multiple fingerprints? Just curious, it seems odd.

Reply:
This is just a representation of the filters used by machineIdentities query filters.
Right now we don't need to query for multiple MI, but in the future, this request could be used to query for both a single MI or multiple MIs (like by keystoreID or by fingerprints)